doc/terminology.rst

```diff
@@ -4,40 +4,105 @@
+        ``DataArray``. In the usual one-dimensional case, the coordinate array's
+        values can loosely be thought of as tick labels along a dimension. There
+        are two types of coordinate arrays: *dimension coordinates* and
+        *non-dimension coordinates* (see below). A coordinate named ``x`` can be
+        retrieved from ``arr.coords[x]``. A ``DataArray`` can have more
+        coordinates than dimensions because a single dimension can be labeled by
+        multiple coordinate arrays. However, only one coordinate array can be a
+        assigned as a particular dimension's dimension coordinate array. As a
+        consequence, ``len(arr.dims) <= len(arr.coords)`` in general.
+
+    Dimension coordinate
+        A one-dimensional coordinate array assigned to ``arr`` with both a name
+        and dimension name in ``arr.dims``. Dimension coordinates are used for
+        label-based indexing and alignment, like the index found on a
+        :py:class:`pandas.DataFrame` or :py:class:`pandas.Series`. In fact,
+        dimension coordinates use :py:class:`pandas.Index` objects under the
+        hood for efficient computation. Dimension coordinates are marked by
+        ``*`` when printing a ``DataArray`` or ``Dataset``.
+
+    Non-dimension coordinate
+        A coordinate array assigned to ``arr`` with a name in ``arr.coords`` but
+        *not* in ``arr.dims``. These coordinates arrays can be one-dimensional
+        or multidimensional, and they are useful for auxiliary labeling. As an
+        example, multidimensional coordinates are often used in geoscience
+        datasets when :doc:`the data's physical coordinates (such as latitude
+        and longitude) differ from their logical coordinates
+        <examples/multidimensional-coords>`. However, non-dimension coordinates
+        are not indexed, and any operation on non-dimension coordinates that
+        leverages indexing will fail. Printing ``arr.coords`` will print all of
+        ``arr``'s coordinate names, with the corresponding dimension(s) in
+        parentheses. For example, ``coord_name (dim_name) 1 2 3 ...``.
+
+    Index
+        An *index* is a data structure optimized for efficient selecting and
+        slicing of an associated array. Xarray creates indexes for dimension
+        coordinates so that operations along dimensions are fast, while
+        non-dimension coordinates are not indexed. Under the hood, indexes are
+        implemented as :py:class:`pandas.Index` objects. The index associated
+        with dimension name ``x`` can be retrieved by ``arr.indexes[x]``. By
+        construction, ``len(arr.dims) == len(arr.indexes)``
+
+    name
+        The names of dimensions, coordinates, DataArray objects and data
+        variables can be anything as long as they are :term:`hashable`. However,
+        it is preferred to use :py:class:`str` typed names.
+
+    scalar
+        By definition, a scalar is not :term:`array_like`. That means that,
```

Comment:
Should "array like" also be defined somewhere in order for a link to work? Or does this just add some sort of formatting?

My definition of a "scalar" might be anything that isn't an array and that when converted into a NumPy array would have 0 dimensions.

NumPy defines "array like" as anything that can be converted into an array. So that definition includes scalar values.

Reply:
right now we're linking to numpy's definition of the term (thanks to `intersphinx`), so defining it on our own might not be necessary.

I updated the definition, is it better now?